Jint/Native/RegExp/RegExpPrototype.cs

```diff
@@ -902,31 +903,51 @@ private static JsValue RegExpBuiltinExec(JsRegExp R, string s)
 
             // the stateful version
             Match match;
-            while (true)
+
+            if (lastIndex > length)
             {
-                if (lastIndex > length)
-                {
-                    R.Set(JsRegExp.PropertyLastIndex, JsNumber.PositiveZero, true);
-                    return Null;
-                }
+                R.Set(JsRegExp.PropertyLastIndex, JsNumber.PositiveZero, true);
+                return Null;
+            }
 
-                match = R.Value.Match(s, (int) lastIndex);
-                var success = match.Success && (!sticky || match.Index == (int) lastIndex);
-                if (!success)
+            var startAt = (int) lastIndex;
+            while (true)
+            {
+                match = R.Value.Match(s, startAt);
+
+                // The conversion of Unicode regex patterns to .NET Regex has some flaws:
+                // when the pattern may match empty strings, the adapted Regex will return empty string matches
+                // in the middle of surrogate pairs. As a best effort solution, we remove these fake positive matches.
+                // (See also: https://github.com/sebastienros/esprima-dotnet/pull/364#issuecomment-1606045259)
+
+                if (match.Success
+                    && fullUnicode
+                    && match.Length == 0
+                    && 0 < match.Index && match.Index < s.Length
+                    && char.IsHighSurrogate(s[match.Index - 1]) && char.IsLowSurrogate(s[match.Index]))
                 {
-                    R.Set(JsRegExp.PropertyLastIndex, JsNumber.PositiveZero, true);
-                    return Null;
+                    startAt++;
+                    continue;
                 }
 
                 break;
             }
 
-            var e = match.Index + match.Length;
-            if (fullUnicode)
+            var success = match.Success && (!sticky || match.Index == (int) lastIndex);
+            if (!success)
             {
-                e = GetStringIndex(s, e);
+                R.Set(JsRegExp.PropertyLastIndex, JsNumber.PositiveZero, true);
+                return Null;
             }
 
+            var e = match.Index + match.Length;
+
+            // .NET regexes doesn't returns code point indices
```

Comment:
maybe can remove? (though this was translated from spec I guess)

Reply:
We can, I was just not sure whether to keep this piece of code around as it's an exact translation of the spec.